HLTrigger/Muon/interface/HLTTriMuonIsolation.h

```diff
@@ -0,0 +1,261 @@
+    MinTriMuonMass_                                                      (iConfig.getParameter<double>       ("MinTriMuonMass"    )) ,
+    MaxTriMuonMass_                                                      (iConfig.getParameter<double>       ("MaxTriMuonMass"    )) , 
+    MaxTriMuonRadius_                                                    (iConfig.getParameter<double>       ("MaxTriMuonRadius"  )) , 
+    TriMuonAbsCharge_                                                    (iConfig.getParameter<int>          ("TriMuonAbsCharge"  )) ,
+    MaxDZ_                                                               (iConfig.getParameter<double>       ("MaxDZ"             )) , 
+    EnableRelIso_                                                        (iConfig.getParameter<bool>         ("EnableRelIso"      )) ,
+    EnableAbsIso_                                                        (iConfig.getParameter<bool>         ("EnableAbsIso"      )) 
+{
+    //register products
+    produces<reco::CompositeCandidateCollection>("Taus");
+    produces<reco::CompositeCandidateCollection>("SelectedTaus");
+}
+
+HLTTriMuonIsolation::~HLTTriMuonIsolation(){ }
+
+void
+HLTTriMuonIsolation::produce(edm::StreamID sid, edm::Event & iEvent, edm::EventSetup const & iSetup) const
+{
+    std::unique_ptr<reco::CompositeCandidateCollection> Taus        (new reco::CompositeCandidateCollection);
+    std::unique_ptr<reco::CompositeCandidateCollection> SelectedTaus(new reco::CompositeCandidateCollection);
+
+    // Get the L3 muon candidates
+    edm::Handle<reco::RecoChargedCandidateCollection> L3MuCands;
+    iEvent.getByToken(L3MuonsToken_, L3MuCands);
+
+    // Get the L3 muon candidates that passed the filter
+    edm::Handle<trigger::TriggerFilterObjectWithRefs> L3DiMuonsFilterCands;
+    iEvent.getByToken(L3DiMuonsFilterToken_, L3DiMuonsFilterCands);
+    
+    std::vector<reco::RecoChargedCandidateRef> PassedL3Muons;
+    L3DiMuonsFilterCands->getObjects(trigger::TriggerMuon, PassedL3Muons);
+  
+    // Get the Trk + L3 muon candidates (after merging)
+    edm::Handle<reco::RecoChargedCandidateCollection> AllMuCands;
+    iEvent.getByToken(AllMuonsToken_, AllMuCands);
+    
+    // Get iso tracks
+    edm::Handle<reco::TrackCollection> IsoTracks;
+    iEvent.getByToken(IsoTracksToken_, IsoTracks);
+          
+    if (AllMuCands->size() >= 3 && L3MuCands->size() >= 2){
+        // Create the 3-muon candidates
+        // loop over L3/Trk muons and create all combinations
+        auto AllMuCands_end = AllMuCands->end();
+        for (auto i = AllMuCands->begin(); i != AllMuCands_end; ++i) {
+            // check that muon_i passes the previous filter
+            bool passingPreviousFilter_1 = false;
+            const reco::TrackRef &tk_i = i->track();
+            for (std::vector<reco::RecoChargedCandidateRef>::const_iterator imu = PassedL3Muons.begin(); imu != PassedL3Muons.end(); ++imu){
```

Comment:
i didn't comment on the outer loop and the next loop because indeed they are not simple (though possible) - i had assumed these second loops can be

for ( auto const & imu  : PassedL3Muons) {
   reco::TrackRef candTrkRef = imu->get<reco::TrackRef>();
....

(but why do you need the trackref at all? btw)